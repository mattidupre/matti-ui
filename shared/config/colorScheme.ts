import { UI_PREFIX } from '../constants';
import { defineColorSchemeConfig } from './lib/defineColorSchemeConfig';

export const COLOR_SCHEME_CONFIG = defineColorSchemeConfig({
  defaultColorScheme: 'dark',
  className: {
    light: `${UI_PREFIX}-color-scheme-light`,
    dark: `${UI_PREFIX}-color-scheme-dark`,
  },
});
