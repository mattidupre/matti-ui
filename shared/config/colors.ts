import { defineColors } from './lib/defineColors';

// https://leonardocolor.io/
// https://uxdesign.cc/creating-contrast-based-themes-with-leonardo-32b6219a090f

export const COLORS_CONFIG = defineColors([
  {
    id: 'site',
    name: 'Site',
    isAdjustable: false,
    base: { hue: 0 },
    swatches: [
      {
        id: 'fill',
        color: {
          light: { lightness: 0 },
          dark: { lightness: 1 },
        },
      },
      {
        id: 'background',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
    ],
  },
  {
    id: 'fill',
    name: 'Fill',
    isAdjustable: false,
    base: { chroma: 0 },
    swatches: [
      {
        id: '100',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '200',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '300',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '400',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '500',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
    ],
  },
  {
    id: 'background',
    name: 'Background',
    isAdjustable: false,
    base: { chroma: 0 },
    swatches: [
      {
        id: '500',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '600',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '700',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '800',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '900',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
    ],
  },
  {
    id: 'primary',
    name: 'Primary',
    isAdjustable: true,
    base: { hue: 0 },
    swatches: [
      {
        id: '100',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '200',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '300',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '400',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '500',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '600',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '700',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '800',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '900',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
    ],
  },
  {
    id: 'accent',
    name: 'Accent',
    isAdjustable: true,
    base: { hue: 120 },
    swatches: [
      {
        id: '100',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '200',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '300',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '400',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '500',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '600',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '700',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '800',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: '900',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
    ],
  },
  {
    id: 'form',
    name: 'Form',
    isAdjustable: false,
    base: { chroma: 0 },
    swatches: [
      {
        id: 'fill',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: 'background',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: 'disabled',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: 'outline',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
      {
        id: 'highlight',
        color: {
          light: { lightness: 1 },
          dark: { lightness: 0 },
        },
      },
    ],
  },
]);
