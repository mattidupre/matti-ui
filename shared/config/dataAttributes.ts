import { defineDataAttributesConfig } from './lib/defineDataAttributesConfig';

export const DATA_ATTRIBUTES_CONFIG = defineDataAttributesConfig({
  flags: {
    isColorSchemeLight: 'data-color-scheme-light',
    isColorSchemeDark: 'data-color-scheme-dark',
    isLoading: 'data-loading',
    isHovered: 'data-hovered',
    isSelected: 'data-selected',
    isPressed: 'data-pressed',
    isFocused: 'data-focused',
    isDisabled: 'data-disabled',
    isEmpty: 'data-empty',
    isOpen: 'data-open',
    isError: 'data-error',
    isChecked: 'data-checked',
    isDragging: 'data-dragging',
  },
});
