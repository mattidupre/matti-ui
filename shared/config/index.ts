export * from './colors';
export * from './colorScheme';
export * from './dataAttributes';
export * from './typography';
