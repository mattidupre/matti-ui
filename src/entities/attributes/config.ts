import type { DataAttributeFlagsConfig } from './entities';

export const DATA_ATTRIBUTE_FLAGS = {
  isColorSchemeLight: 'data-color-scheme-light',
  isColorSchemeDark: 'data-color-scheme-dark',
  isLoading: 'data-loading',
  isHovered: 'data-hovered',
  isSelected: 'data-selected',
  isPressed: 'data-pressed',
  isFocused: 'data-focused',
  isDisabled: 'data-disabled',
  isEmpty: 'data-empty',
  isOpen: 'data-open',
  isError: 'data-error',
  isChecked: 'data-checked',
  isDragging: 'data-dragging',
} as const satisfies DataAttributeFlagsConfig;
