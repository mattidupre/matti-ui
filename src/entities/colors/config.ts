import { definePalettes } from './definePalettes';

export const BASE_PALETTES = definePalettes({
  site: {
    fill: { light: { lightness: 0 }, dark: { lightness: 1 } },
    background: { light: { lightness: 1 }, dark: { lightness: 0 } },
    browser: {
      /* extends: 'primary.500' */
    },
  },
  fill: {
    '100': { light: { lightness: 1 } },
    '200': {},
    '300': {},
    '400': {},
    '500': {},
  },
  background: {
    '500': {},
    '600': {},
    '700': {},
    '800': {},
    '900': {},
  },
  primary: {
    _base: { hue: 0 },
    '100': {},
    '200': {},
    '300': {},
    '400': {},
    '500': {},
    '600': {},
    '700': {},
    '800': {},
    '900': {},
  },
  accent: {
    _base: { hue: 120 },
    '100': {},
    '200': {},
    '300': {},
    '400': {},
    '500': {},
    '600': {},
    '700': {},
    '800': {},
    '900': {},
  },
  form: {
    fill: {},
    background: {},
    disabled: {},
    outline: {},
    highlight: {},
  },
});
