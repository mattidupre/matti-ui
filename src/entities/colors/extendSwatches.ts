import { mapValues } from 'lodash-es';
import type { ColorsOptions } from './types';
import type { PalettesConfig } from './definePalettes';

export const extendSwatches = <TOptions extends ColorsOptions>(
  palettes: PalettesConfig,
  colors: TOptions,
): Pick<PalettesConfig['swatches'], keyof TOptions> => {
  const extendedPalettes = {} as Pick<
    PalettesConfig['swatches'],
    keyof TOptions
  >;

  for (const paletteId of Object.keys(colors) as ReadonlyArray<
    keyof typeof colors
  >) {
    if (!colors[paletteId]) {
      continue;
    }
    const {
      light: lightOptions,
      dark: darkOptions,
      ...baseOptions
    } = colors[paletteId];
    const swatches = mapValues(
      palettes[paletteId].swatches,
      ({ light, dark }) => ({
        light: {
          ...light,
          ...baseOptions,
          ...lightOptions,
        },
        dark: {
          ...dark,
          ...baseOptions,
          ...darkOptions,
        },
      }),
    );
    extendedPalettes[paletteId] = swatches;
  }
  return extendedPalettes;
};
