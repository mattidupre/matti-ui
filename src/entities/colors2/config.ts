import type { PalettesOptions } from './entities';
import { defineColorsConfig } from './lib/defineColorsConfig';

export const PALETTES_OPTIONS_DEFAULT: PalettesOptions = {
  site: {
    base: { hue: 0 },
    swatches: {
      fill: { light: { lightness: 0 }, dark: { lightness: 1 } },
      background: { light: { lightness: 1 }, dark: { lightness: 0 } },
      browser: {
        /* extends: 'primary.500' */
      },
    },
  },
  fill: {
    base: { chroma: 0 },
    swatches: {
      '100': { light: { lightness: 1 } },
      '200': {},
      '300': {},
      '400': {},
      '500': {},
    },
  },
  background: {
    base: { chroma: 0 },
    swatches: {
      '500': {},
      '600': {},
      '700': {},
      '800': {},
      '900': {},
    },
  },
  primary: {
    base: { hue: 0 },
    swatches: {
      '100': {},
      '200': {},
      '300': {},
      '400': {},
      '500': {},
      '600': {},
      '700': {},
      '800': {},
      '900': {},
    },
  },
  accent: {
    base: { hue: 120 },
    swatches: {
      '100': {},
      '200': {},
      '300': {},
      '400': {},
      '500': {},
      '600': {},
      '700': {},
      '800': {},
      '900': {},
    },
  },
  form: {
    base: { hue: 0, chroma: 0 },
    swatches: {
      fill: {},
      background: {},
      disabled: {},
      outline: {},
      highlight: {},
    },
  },
};

export const PALETTES_CONFIG = defineColorsConfig(PALETTES_OPTIONS_DEFAULT);
