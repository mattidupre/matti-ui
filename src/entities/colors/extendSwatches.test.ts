import { describe, test, expect } from 'vitest';
import { extendSwatches } from './extendSwatches';
import { BASE_PALETTES } from './config';

describe(extendSwatches, () => {
  test('empty swatches', () => {
    expect(extendSwatches(BASE_PALETTES, {})).toStrictEqual({});
  });

  test('single swatch', () => {
    const RESULT = extendSwatches(BASE_PALETTES, {
      primary: {
        hue: 123,
      },
    });

    expect(RESULT).toMatchSnapshot();
    expect(RESULT.primary[500].light.hue).toBe(123);
    expect(RESULT.primary[500].dark.hue).toBe(123);
  });

  test('multiple swatches', () => {
    const RESULT = extendSwatches(BASE_PALETTES, {
      primary: {
        hue: 123,
      },
      accent: {
        hue: 321,
      },
    });

    expect(RESULT).toMatchSnapshot();
    expect(RESULT.primary[500].light.hue).toBe(123);
    expect(RESULT.primary[500].dark.hue).toBe(123);
    expect(RESULT.accent[500].light.hue).toBe(321);
    expect(RESULT.accent[500].dark.hue).toBe(321);
  });
});
