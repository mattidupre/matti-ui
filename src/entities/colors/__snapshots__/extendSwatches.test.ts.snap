// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extendSwatches > multiple swatches 1`] = `
{
  "accent": {
    "100": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "200": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "300": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "400": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "500": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "600": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "700": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "800": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
    "900": {
      "dark": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 321,
        "lightness": 0,
      },
    },
  },
  "primary": {
    "100": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "200": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "300": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "400": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "500": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "600": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "700": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "800": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "900": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
  },
}
`;

exports[`extendSwatches > single swatch 1`] = `
{
  "primary": {
    "100": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "200": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "300": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "400": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "500": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "600": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "700": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "800": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
    "900": {
      "dark": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
      "light": {
        "chroma": 0,
        "hue": 123,
        "lightness": 0,
      },
    },
  },
}
`;
